app/inpututils.cpp

```diff
@@ -2222,3 +2222,65 @@ double InputUtils::pixelDistanceBetween( const QPointF &p1, const QPointF &p2 )
 {
   return std::hypot( p1.x() - p2.x(), p1.y() - p2.y() );
 }
+
+bool InputUtils::layerHasGeometry( const QgsVectorLayer *layer )
+{
+  if ( !layer || !layer->isValid() )
+    return false;
+  return layer->wkbType() != Qgis::WkbType::NoGeometry && layer->wkbType() != Qgis::WkbType::Unknown;
+}
+
+bool InputUtils::layerVisible( QgsMapLayer *layer )
+{
+  if ( !layer || !layer->isValid() )
+    return false;
+
+  QgsLayerTree *root = QgsProject::instance()->layerTreeRoot();
```

Comment:
We should not use `QgsProject::instance()` --> pass it in to the function instead